In [32]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [33]:
df = pd.read_csv("garments_worker_productivity.csv")

In [34]:
df

,date,quarter,department,day,team,targeted_productivity,smv,wip,over_time,incentive,idle_time,idle_men,no_of_style_change,no_of_workers,actual_productivity
0,1/1/2015,Quarter1,sweing,Thursday,8,0.80,26.16,1108.0,7080,98,0.0,0,0,59.0,0.940725
1,1/1/2015,Quarter1,finishing,Thursday,1,0.75,3.94,NaN,960,0,0.0,0,0,8.0,0.886500
2,1/1/2015,Quarter1,sweing,Thursday,11,0.80,11.41,968.0,3660,50,0.0,0,0,30.5,0.800570
3,1/1/2015,Quarter1,sweing,Thursday,12,0.80,11.41,968.0,3660,50,0.0,0,0,30.5,0.800570
4,1/1/2015,Quarter1,sweing,Thursday,6,0.80,25.90,1170.0,1920,50,0.0,0,0,56.0,0.800382
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1192,3/11/2015,Quarter2,finishing,Wednesday,10,0.75,2.90,NaN,960,0,0.0,0,0,8.0,0.628333
1193,3/11/2015,Quarter2,finishing,Wednesday,8,0.70,3.90,NaN,960,0,0.0,0,0,8.0,0.625625
1194,3/11/2015,Quarter2,finishing,Wednesday,7,0.65,3.90,NaN,960,0,0.0,0,0,8.0,0.625625
1195,3/11/2015,Quarter2,finishing,Wednesday,9,0.75,2.90,NaN,1800,0,0.0,0,0,15.0,0.505889


In [35]:
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 1197 entries, 0 to 1196
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   date                   1197 non-null   str    
 1   quarter                1197 non-null   str    
 2   department             1197 non-null   str    
 3   day                    1197 non-null   str    
 4   team                   1197 non-null   int64  
 5   targeted_productivity  1197 non-null   float64
 6   smv                    1197 non-null   float64
 7   wip                    691 non-null    float64
 8   over_time              1197 non-null   int64  
 9   incentive              1197 non-null   int64  
 10  idle_time              1197 non-null   float64
 11  idle_men               1197 non-null   int64  
 12  no_of_style_change     1197 non-null   int64  
 13  no_of_workers          1197 non-null   float64
 14  actual_productivity    1197 non-null   float64
dtypes: float64(6), 

In [36]:
print(df.isnull().sum())

date                       0
quarter                    0
department                 0
day                        0
team                       0
targeted_productivity      0
smv                        0
wip                      506
over_time                  0
incentive                  0
idle_time                  0
idle_men                   0
no_of_style_change         0
no_of_workers              0
actual_productivity        0
dtype: int64


In [37]:
for col in ["date", "quarter", "department", "day"]:
    print("\n", col)
    print(df[col].unique())


 date
<ArrowStringArray>
[ '1/1/2015',  '1/3/2015',  '1/4/2015',  '1/5/2015',  '1/6/2015',  '1/7/2015',
  '1/8/2015', '1/10/2015', '1/11/2015', '1/12/2015', '1/13/2015', '1/14/2015',
 '1/15/2015', '1/17/2015', '1/18/2015', '1/19/2015', '1/20/2015', '1/21/2015',
 '1/22/2015', '1/24/2015', '1/25/2015', '1/26/2015', '1/27/2015', '1/28/2015',
 '1/29/2015', '1/31/2015',  '2/1/2015',  '2/2/2015',  '2/3/2015',  '2/4/2015',
  '2/5/2015',  '2/7/2015',  '2/8/2015',  '2/9/2015', '2/10/2015', '2/11/2015',
 '2/12/2015', '2/14/2015', '2/15/2015', '2/16/2015', '2/17/2015', '2/18/2015',
 '2/19/2015', '2/22/2015', '2/23/2015', '2/24/2015', '2/25/2015', '2/26/2015',
 '2/28/2015',  '3/1/2015',  '3/2/2015',  '3/3/2015',  '3/4/2015',  '3/5/2015',
  '3/7/2015',  '3/8/2015',  '3/9/2015', '3/10/2015', '3/11/2015']
Length: 59, dtype: str

 quarter
<ArrowStringArray>
['Quarter1', 'Quarter2', 'Quarter3', 'Quarter4', 'Quarter5']
Length: 5, dtype: str

 department
<ArrowStringArray>
['sweing', 'finishing ', 'fini

In [38]:
print(df.describe().T)

                        count         mean          std       min  \
team                   1197.0     6.426901     3.463963  1.000000   
targeted_productivity  1197.0     0.729632     0.097891  0.070000   
smv                    1197.0    15.062172    10.943219  2.900000   
wip                     691.0  1190.465991  1837.455001  7.000000   
over_time              1197.0  4567.460317  3348.823563  0.000000   
incentive              1197.0    38.210526   160.182643  0.000000   
idle_time              1197.0     0.730159    12.709757  0.000000   
idle_men               1197.0     0.369256     3.268987  0.000000   
no_of_style_change     1197.0     0.150376     0.427848  0.000000   
no_of_workers          1197.0    34.609858    22.197687  2.000000   
actual_productivity    1197.0     0.735091     0.174488  0.233705   

                               25%          50%          75%           max  
team                      3.000000     6.000000     9.000000     12.000000  
targeted_producti

In [39]:
print(df.groupby("department")["wip"].apply(lambda x: x.isnull().sum()))

print("\nMissing WIP percentage:")
print(df["wip"].isnull().mean() * 100)

department
finishing     249
finishing     257
sweing          0
Name: wip, dtype: int64

Missing WIP percentage:
42.272347535505425


In [40]:
df["department"] = df["department"].str.strip()

print(df.groupby("department")["wip"].agg(["count", "median", "mean"]))

            count  median         mean
department                            
finishing       0     NaN          NaN
sweing        691  1039.0  1190.465991


In [41]:
print(df["department"].unique())
print("\nCounts:")
print(df["department"].value_counts())

<ArrowStringArray>
['sweing', 'finishing']
Length: 2, dtype: str

Counts:
department
sweing       691
finishing    506
Name: count, dtype: int64


In [42]:
print("First date:", df["date"].iloc[0])
print("Last date:", df["date"].iloc[-1])
print("\nNumber of unique dates:", df["date"].nunique())

First date: 1/1/2015
Last date: 3/11/2015

Number of unique dates: 59


In [43]:
meets_target = df["actual_productivity"] >= df["targeted_productivity"]
print(meets_target.value_counts())
print("\nPercentage:")
print(meets_target.value_counts(normalize=True) * 100)

True     875
False    322
Name: count, dtype: int64

Percentage:
True     73.099415
False    26.900585
Name: proportion, dtype: float64


In [44]:
print(df.corr(numeric_only=True)["actual_productivity"].sort_values(ascending=False))

actual_productivity      1.000000
targeted_productivity    0.421594
wip                      0.131147
incentive                0.076538
over_time               -0.054206
no_of_workers           -0.057991
idle_time               -0.080851
smv                     -0.122089
team                    -0.148753
idle_men                -0.181734
no_of_style_change      -0.207366
Name: actual_productivity, dtype: float64


In [45]:
for col in ["quarter", "department", "day"]:
    print("\n" + col)
    print(df[col].value_counts())


quarter
quarter
Quarter1    360
Quarter2    335
Quarter4    248
Quarter3    210
Quarter5     44
Name: count, dtype: int64

department
department
sweing       691
finishing    506
Name: count, dtype: int64

day
day
Wednesday    208
Sunday       203
Tuesday      201
Thursday     199
Monday       199
Saturday     187
Name: count, dtype: int64


In [46]:
print(df["date"].head(10))
print("\nNumber of unique dates:", df["date"].nunique())

0    1/1/2015
1    1/1/2015
2    1/1/2015
3    1/1/2015
4    1/1/2015
5    1/1/2015
6    1/1/2015
7    1/1/2015
8    1/1/2015
9    1/1/2015
Name: date, dtype: str

Number of unique dates: 59


In [47]:
date_check = pd.to_datetime(df["date"])

print("Start date:", date_check.min())
print("End date:", date_check.max())
print("\nMonths:")
print(date_check.dt.month.value_counts().sort_index())

Start date: 2015-01-01 00:00:00
End date: 2015-03-11 00:00:00

Months:
date
1    542
2    443
3    212
Name: count, dtype: int64


In [48]:
print("Regression target:")
print(df["actual_productivity"].describe())

print("\nClassification target:")
meets_target = (df["actual_productivity"] >= df["targeted_productivity"]).astype(int)
print(meets_target.value_counts())
print("\nPercentage:")
print(meets_target.value_counts(normalize=True) * 100)

Regression target:
count    1197.000000
mean        0.735091
std         0.174488
min         0.233705
25%         0.650307
50%         0.773333
75%         0.850253
max         1.120437
Name: actual_productivity, dtype: float64

Classification target:
1    875
0    322
Name: count, dtype: int64

Percentage:
1    73.099415
0    26.900585
Name: proportion, dtype: float64


In [49]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns

for col in numeric_cols:
    Q1 = df[col].quantile(0.25)
    Q3 = df[col].quantile(0.75)
    IQR = Q3 - Q1

    lower = Q1 - 1.5 * IQR
    upper = Q3 + 1.5 * IQR

    outliers = ((df[col] < lower) | (df[col] > upper)).sum()

    print(col, "->", outliers)

team -> 0
targeted_productivity -> 79
smv -> 0
wip -> 22
over_time -> 1
incentive -> 11
idle_time -> 18
idle_men -> 18
no_of_style_change -> 147
no_of_workers -> 0
actual_productivity -> 54


In [50]:
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns
print(df[numeric_cols].describe().T[["min", "max", "mean", "std"]])

                            min           max         mean          std
team                   1.000000     12.000000     6.426901     3.463963
targeted_productivity  0.070000      0.800000     0.729632     0.097891
smv                    2.900000     54.560000    15.062172    10.943219
wip                    7.000000  23122.000000  1190.465991  1837.455001
over_time              0.000000  25920.000000  4567.460317  3348.823563
incentive              0.000000   3600.000000    38.210526   160.182643
idle_time              0.000000    300.000000     0.730159    12.709757
idle_men               0.000000     45.000000     0.369256     3.268987
no_of_style_change     0.000000      2.000000     0.150376     0.427848
no_of_workers          2.000000     89.000000    34.609858    22.197687
actual_productivity    0.233705      1.120437     0.735091     0.174488


In [51]:
corr = df[numeric_cols].corr()

print(corr.round(2))

                       team  targeted_productivity   smv   wip  over_time  \
team                   1.00                   0.03 -0.11 -0.03      -0.10   
targeted_productivity  0.03                   1.00 -0.07  0.06      -0.09   
smv                   -0.11                  -0.07  1.00 -0.04       0.67   
wip                   -0.03                   0.06 -0.04  1.00       0.02   
over_time             -0.10                  -0.09  0.67  0.02       1.00   
incentive             -0.01                   0.03  0.03  0.17      -0.00   
idle_time              0.00                  -0.06  0.06 -0.03       0.03   
idle_men               0.03                  -0.05  0.11 -0.05      -0.02   
no_of_style_change    -0.01                  -0.21  0.32 -0.07       0.06   
no_of_workers         -0.08                  -0.08  0.91  0.03       0.73   
actual_productivity   -0.15                   0.42 -0.12  0.13      -0.05   

                       incentive  idle_time  idle_men  no_of_style_change  

In [52]:
df["MeetsTarget"] = (
    df["actual_productivity"] >= df["targeted_productivity"]
).astype(int)

print(df[["targeted_productivity", "actual_productivity", "MeetsTarget"]].head(10))

   targeted_productivity  actual_productivity  MeetsTarget
0                   0.80             0.940725            1
1                   0.75             0.886500            1
2                   0.80             0.800570            1
3                   0.80             0.800570            1
4                   0.80             0.800382            1
5                   0.80             0.800125            1
6                   0.75             0.755167            1
7                   0.75             0.753683            1
8                   0.75             0.753098            1
9                   0.75             0.750428            1


In [53]:
from sklearn.model_selection import train_test_split

reg_target = df["actual_productivity"]
clf_target = df["MeetsTarget"]

train_idx, test_idx = train_test_split(
    df.index,
    test_size=0.2,
    random_state=42
)

print("Training rows:", len(train_idx))
print("Testing rows:", len(test_idx))

Training rows: 957
Testing rows: 240


In [54]:
features = [
    "quarter",
    "department",
    "day",
    "team",
    "targeted_productivity",
    "smv",
    "wip",
    "over_time",
    "incentive",
    "idle_time",
    "idle_men",
    "no_of_style_change",
    "no_of_workers"
]

X = df[features].copy()

print("Shape:", X.shape)
print(X.head())

Shape: (1197, 13)
    quarter department       day  team  targeted_productivity    smv     wip  \
0  Quarter1     sweing  Thursday     8                   0.80  26.16  1108.0   
1  Quarter1  finishing  Thursday     1                   0.75   3.94     NaN   
2  Quarter1     sweing  Thursday    11                   0.80  11.41   968.0   
3  Quarter1     sweing  Thursday    12                   0.80  11.41   968.0   
4  Quarter1     sweing  Thursday     6                   0.80  25.90  1170.0   

   over_time  incentive  idle_time  idle_men  no_of_style_change  \
0       7080         98        0.0         0                   0   
1        960          0        0.0         0                   0   
2       3660         50        0.0         0                   0   
3       3660         50        0.0         0                   0   
4       1920         50        0.0         0                   0   

   no_of_workers  
0           59.0  
1            8.0  
2           30.5  
3           30.5

In [55]:
X["month"] = pd.to_datetime(df["date"]).dt.month

print(X[["month"]].head())
print("\nMonth counts:")
print(X["month"].value_counts().sort_index())

   month
0      1
1      1
2      1
3      1
4      1

Month counts:
month
1    542
2    443
3    212
Name: count, dtype: int64


In [56]:
X = X.drop(columns=["date"], errors="ignore")

print("Features:", X.columns.tolist())
print("Shape:", X.shape)

Features: ['quarter', 'department', 'day', 'team', 'targeted_productivity', 'smv', 'wip', 'over_time', 'incentive', 'idle_time', 'idle_men', 'no_of_style_change', 'no_of_workers', 'month']
Shape: (1197, 14)


In [57]:
wip_median = X.loc[train_idx, "wip"].median()

print("Training WIP median:", wip_median)

Training WIP median: 1023.0


In [58]:
X["wip_missing"] = X["wip"].isnull().astype(int)

print(X[["wip", "wip_missing"]].head(10))
print("\nMissing indicator counts:")
print(X["wip_missing"].value_counts())

      wip  wip_missing
0  1108.0            0
1     NaN            1
2   968.0            0
3   968.0            0
4  1170.0            0
5   984.0            0
6     NaN            1
7   795.0            0
8   733.0            0
9   681.0            0

Missing indicator counts:
wip_missing
0    691
1    506
Name: count, dtype: int64


In [59]:
X["wip"] = X["wip"].fillna(wip_median)

print("Missing WIP after imputation:", X["wip"].isnull().sum())
print(X[["wip", "wip_missing"]].head(10))

Missing WIP after imputation: 0
      wip  wip_missing
0  1108.0            0
1  1023.0            1
2   968.0            0
3   968.0            0
4  1170.0            0
5   984.0            0
6  1023.0            1
7   795.0            0
8   733.0            0
9   681.0            0


In [60]:
X["wip"] = X["wip"].fillna(wip_median)

print("Missing WIP after imputation:", X["wip"].isnull().sum())
print(X[["wip", "wip_missing"]].head(10))

Missing WIP after imputation: 0
      wip  wip_missing
0  1108.0            0
1  1023.0            1
2   968.0            0
3   968.0            0
4  1170.0            0
5   984.0            0
6  1023.0            1
7   795.0            0
8   733.0            0
9   681.0            0


In [61]:
print("Number of features:", X.shape[1])
print("\nCategorical columns:")
print(["quarter", "department", "day", "team"])

print("\nNumerical columns:")
print([
    "targeted_productivity",
    "smv",
    "wip",
    "over_time",
    "incentive",
    "idle_time",
    "idle_men",
    "no_of_style_change",
    "no_of_workers",
    "month",
    "wip_missing"
])

Number of features: 15

Categorical columns:
['quarter', 'department', 'day', 'team']

Numerical columns:
['targeted_productivity', 'smv', 'wip', 'over_time', 'incentive', 'idle_time', 'idle_men', 'no_of_style_change', 'no_of_workers', 'month', 'wip_missing']


In [62]:
categorical_cols = ["quarter", "department", "day", "team"]

numerical_cols = [
    "targeted_productivity",
    "smv",
    "wip",
    "over_time",
    "incentive",
    "idle_time",
    "idle_men",
    "no_of_style_change",
    "no_of_workers",
    "month",
    "wip_missing"
]

print("Categorical:", categorical_cols)
print("Numerical:", numerical_cols)

Categorical: ['quarter', 'department', 'day', 'team']
Numerical: ['targeted_productivity', 'smv', 'wip', 'over_time', 'incentive', 'idle_time', 'idle_men', 'no_of_style_change', 'no_of_workers', 'month', 'wip_missing']


In [63]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numerical_cols),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_cols)
    ]
)

print("Preprocessor created")

Preprocessor created


In [64]:
X_train = X.loc[train_idx]
X_test = X.loc[test_idx]

preprocessor.fit(X_train)

print("Preprocessor fitted on training data")

Preprocessor fitted on training data


In [65]:
X_train_processed = preprocessor.transform(X_train)
X_test_processed = preprocessor.transform(X_test)

print("Training shape:", X_train_processed.shape)
print("Testing shape:", X_test_processed.shape)

Training shape: (957, 36)
Testing shape: (240, 36)


In [66]:
y_reg_train = reg_target.loc[train_idx]
y_reg_test = reg_target.loc[test_idx]

y_clf_train = clf_target.loc[train_idx]
y_clf_test = clf_target.loc[test_idx]

print("Regression:", len(y_reg_train), len(y_reg_test))
print("Classification:", len(y_clf_train), len(y_clf_test))

Regression: 957 240
Classification: 957 240


In [67]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import time

In [68]:
reg_model = LinearRegression()

start_time = time.perf_counter()

reg_model.fit(X_train_processed, y_reg_train)

train_time_reg = time.perf_counter() - start_time

print("Training time:", train_time_reg, "seconds")

Training time: 0.014328799999930197 seconds


In [69]:
start_time = time.perf_counter()

y_reg_pred = reg_model.predict(X_test_processed)

predict_time_reg = time.perf_counter() - start_time

print("Prediction time:", predict_time_reg, "seconds")
print("\nFirst 10 predictions:")
print(y_reg_pred[:10])

Prediction time: 0.002125700000760844 seconds

First 10 predictions:
[0.63679562 0.73964249 0.81039749 0.72848321 0.76797792 0.79974827
 0.58968705 0.62668732 0.65751647 0.71050959]


In [70]:
mae_reg = mean_absolute_error(y_reg_test, y_reg_pred)
rmse_reg = mean_squared_error(y_reg_test, y_reg_pred) ** 0.5
r2_reg = r2_score(y_reg_test, y_reg_pred)

print("MAE :", mae_reg)
print("RMSE:", rmse_reg)
print("R2  :", r2_reg)

MAE : 0.11114585853975874
RMSE: 0.14935052152028525
R2  : 0.159943267095781


In [71]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [72]:
clf_model = LogisticRegression(max_iter=1000)

start_time = time.perf_counter()

clf_model.fit(X_train_processed, y_clf_train)

train_time_clf = time.perf_counter() - start_time

print("Training time:", train_time_clf, "seconds")

Training time: 0.023234899999806657 seconds


In [73]:
start_time = time.perf_counter()

y_clf_pred = clf_model.predict(X_test_processed)

predict_time_clf = time.perf_counter() - start_time

print("Prediction time:", predict_time_clf, "seconds")
print("\nFirst 20 predictions:")
print(y_clf_pred[:20])

Prediction time: 0.0005437999989226228 seconds

First 20 predictions:
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]


In [74]:
accuracy_clf = accuracy_score(y_clf_test, y_clf_pred)
precision_clf = precision_score(y_clf_test, y_clf_pred)
recall_clf = recall_score(y_clf_test, y_clf_pred)
f1_clf = f1_score(y_clf_test, y_clf_pred)

print("Accuracy :", accuracy_clf)
print("Precision:", precision_clf)
print("Recall   :", recall_clf)
print("F1 Score :", f1_clf)

Accuracy : 0.7666666666666667
Precision: 0.7922705314009661
Recall   : 0.9265536723163842
F1 Score : 0.8541666666666666


In [75]:
sklearn_results = {
    "Regression": {
        "MAE": mae_reg,
        "RMSE": rmse_reg,
        "R2": r2_reg,
        "Train Time": train_time_reg,
        "Prediction Time": predict_time_reg
    },
    "Classification": {
        "Accuracy": accuracy_clf,
        "Precision": precision_clf,
        "Recall": recall_clf,
        "F1": f1_clf,
        "Train Time": train_time_clf,
        "Prediction Time": predict_time_clf
    }
}

print(sklearn_results)

{'Regression': {'MAE': 0.11114585853975874, 'RMSE': 0.14935052152028525, 'R2': 0.159943267095781, 'Train Time': 0.014328799999930197, 'Prediction Time': 0.002125700000760844}, 'Classification': {'Accuracy': 0.7666666666666667, 'Precision': 0.7922705314009661, 'Recall': 0.9265536723163842, 'F1': 0.8541666666666666, 'Train Time': 0.023234899999806657, 'Prediction Time': 0.0005437999989226228}}


In [76]:
X_manual = X.copy()

X_train_manual = X_manual.loc[train_idx].copy()
X_test_manual = X_manual.loc[test_idx].copy()

print("Training shape:", X_train_manual.shape)
print("Testing shape:", X_test_manual.shape)

Training shape: (957, 15)
Testing shape: (240, 15)


In [77]:
categorical_cols_manual = ["quarter", "department", "day", "team"]

numerical_cols_manual = [
    "targeted_productivity",
    "smv",
    "wip",
    "over_time",
    "incentive",
    "idle_time",
    "idle_men",
    "no_of_style_change",
    "no_of_workers",
    "month",
    "wip_missing"
]

print("Categorical columns:", categorical_cols_manual)
print("Numerical columns:", numerical_cols_manual)

Categorical columns: ['quarter', 'department', 'day', 'team']
Numerical columns: ['targeted_productivity', 'smv', 'wip', 'over_time', 'incentive', 'idle_time', 'idle_men', 'no_of_style_change', 'no_of_workers', 'month', 'wip_missing']


In [78]:
X_train_cat = pd.get_dummies(
    X_train_manual[categorical_cols_manual],
    dtype=int
)

X_test_cat = pd.get_dummies(
    X_test_manual[categorical_cols_manual],
    dtype=int
)

X_test_cat = X_test_cat.reindex(
    columns=X_train_cat.columns,
    fill_value=0
)

print("Training categorical shape:", X_train_cat.shape)
print("Testing categorical shape:", X_test_cat.shape)
print("\nEncoded columns:")
print(X_train_cat.columns.tolist())

Training categorical shape: (957, 14)
Testing categorical shape: (240, 14)

Encoded columns:
['team', 'quarter_Quarter1', 'quarter_Quarter2', 'quarter_Quarter3', 'quarter_Quarter4', 'quarter_Quarter5', 'department_finishing', 'department_sweing', 'day_Monday', 'day_Saturday', 'day_Sunday', 'day_Thursday', 'day_Tuesday', 'day_Wednesday']


In [79]:
train_mean = X_train_manual[numerical_cols_manual].mean()
train_std = X_train_manual[numerical_cols_manual].std()

print("Training means:")
print(train_mean)

print("\nTraining standard deviations:")
print(train_std)

Training means:
targeted_productivity       0.727921
smv                        14.847450
wip                      1116.699060
over_time                4482.842215
incentive                  39.882968
idle_time                   0.311912
idle_men                    0.350052
no_of_style_change          0.155695
no_of_workers              33.871473
month                       1.742947
wip_missing                 0.433647
dtype: float64

Training standard deviations:
targeted_productivity       0.100851
smv                        10.946357
wip                      1443.850806
over_time                3353.050392
incentive                 175.879871
idle_time                   5.678645
idle_men                    3.086570
no_of_style_change          0.428827
no_of_workers              22.175689
month                       0.753489
wip_missing                 0.495837
dtype: float64


In [80]:
X_train_num = (
    X_train_manual[numerical_cols_manual] - train_mean
) / train_std

X_test_num = (
    X_test_manual[numerical_cols_manual] - train_mean
) / train_std

print("Training numerical shape:", X_train_num.shape)
print("Testing numerical shape:", X_test_num.shape)

print("\nFirst 5 standardized training rows:")
print(X_train_num.head())

Training numerical shape: (957, 11)
Testing numerical shape: (240, 11)

First 5 standardized training rows:
      targeted_productivity       smv       wip  over_time  incentive  \
1189              -0.276850  1.428105 -0.140388   0.702989  -0.056192   
575                0.218932 -0.996446 -0.064895  -0.656967  -0.226763   
76                 0.218932 -1.091454 -0.064895  -1.050638  -0.226763   
731               -0.276850 -0.977261 -0.064895  -0.800120  -0.226763   
138                0.714713 -0.295756 -0.393877   3.172382   0.131436   

      idle_time  idle_men  no_of_style_change  no_of_workers     month  \
1189  -0.054927 -0.113411            1.968870       1.042968  1.668310   
575   -0.054927 -0.113411           -0.363071      -0.670621  0.341151   
76    -0.054927 -0.113411           -0.363071      -1.166659 -0.986009   
731   -0.054927 -0.113411           -0.363071      -0.850998  0.341151   
138   -0.054927 -0.113411           -0.363071      -0.106940 -0.986009   

      wi

In [81]:
X_train_processed_manual = pd.concat(
    [X_train_num, X_train_cat],
    axis=1
)

X_test_processed_manual = pd.concat(
    [X_test_num, X_test_cat],
    axis=1
)

print("Training shape:", X_train_processed_manual.shape)
print("Testing shape:", X_test_processed_manual.shape)

print("\nFirst 5 rows:")
print(X_train_processed_manual.head())

Training shape: (957, 25)
Testing shape: (240, 25)

First 5 rows:
      targeted_productivity       smv       wip  over_time  incentive  \
1189              -0.276850  1.428105 -0.140388   0.702989  -0.056192   
575                0.218932 -0.996446 -0.064895  -0.656967  -0.226763   
76                 0.218932 -1.091454 -0.064895  -1.050638  -0.226763   
731               -0.276850 -0.977261 -0.064895  -0.800120  -0.226763   
138                0.714713 -0.295756 -0.393877   3.172382   0.131436   

      idle_time  idle_men  no_of_style_change  no_of_workers     month  ...  \
1189  -0.054927 -0.113411            1.968870       1.042968  1.668310  ...   
575   -0.054927 -0.113411           -0.363071      -0.670621  0.341151  ...   
76    -0.054927 -0.113411           -0.363071      -1.166659 -0.986009  ...   
731   -0.054927 -0.113411           -0.363071      -0.850998  0.341151  ...   
138   -0.054927 -0.113411           -0.363071      -0.106940 -0.986009  ...   

      quarter_Quarte

In [82]:
y_train_manual = y_reg_train.to_numpy()
y_test_manual = y_reg_test.to_numpy()

print("Training target shape:", y_train_manual.shape)
print("Testing target shape:", y_test_manual.shape)

print("\nFirst 10 training targets:")
print(y_train_manual[:10])

Training target shape: (957,)
Testing target shape: (240,)

First 10 training targets:
[0.70050526 0.50459649 0.763375   1.09663333 0.80010714 0.6000627
 0.60741667 0.7502069  0.89366319 0.77979167]


In [83]:
X_train_np = X_train_processed_manual.to_numpy()
X_test_np = X_test_processed_manual.to_numpy()

print("Training X shape:", X_train_np.shape)
print("Testing X shape:", X_test_np.shape)

Training X shape: (957, 25)
Testing X shape: (240, 25)


In [84]:
X_train_bias = np.c_[
    np.ones(X_train_np.shape[0]),
    X_train_np
]

X_test_bias = np.c_[
    np.ones(X_test_np.shape[0]),
    X_test_np
]

print("Training shape:", X_train_bias.shape)
print("Testing shape:", X_test_bias.shape)

Training shape: (957, 26)
Testing shape: (240, 26)


In [85]:
start_time = time.perf_counter()

beta = np.linalg.inv(
    X_train_bias.T @ X_train_bias
) @ X_train_bias.T @ y_train_manual

train_time_manual_reg = time.perf_counter() - start_time

print("Number of coefficients:", len(beta))
print("Training time:", train_time_manual_reg, "seconds")
print("\nFirst 10 coefficients:")
print(beta[:10])

Number of coefficients: 26
Training time: 0.008329700000103912 seconds

First 10 coefficients:
[-4.10518038  0.14747677  0.09182602 -0.04178885  0.07257908 -0.04358862
  0.07615808 -0.1241348   0.06030518  0.14709982]


In [86]:
start_time = time.perf_counter()

y_manual_pred = X_test_bias @ beta

predict_time_manual_reg = time.perf_counter() - start_time

print("Prediction time:", predict_time_manual_reg, "seconds")

print("\nFirst 10 predictions:")
print(y_manual_pred[:10])

Prediction time: 0.00010999999904015567 seconds

First 10 predictions:
[-2.30960674  4.6257286   1.02415484 -2.5447968   1.05329968  4.89904673
 -3.42768983 -2.38935318 -1.87703484  0.50675491]


In [87]:
X_train_cat = X_train_manual[categorical_cols_manual].copy()
X_test_cat = X_test_manual[categorical_cols_manual].copy()

X_train_cat["team"] = X_train_cat["team"].astype(str)
X_test_cat["team"] = X_test_cat["team"].astype(str)

X_train_cat = pd.get_dummies(
    X_train_cat,
    dtype=int
)

X_test_cat = pd.get_dummies(
    X_test_cat,
    dtype=int
)

X_test_cat = X_test_cat.reindex(
    columns=X_train_cat.columns,
    fill_value=0
)

print("Training categorical shape:", X_train_cat.shape)
print("Testing categorical shape:", X_test_cat.shape)

Training categorical shape: (957, 25)
Testing categorical shape: (240, 25)


In [88]:
X_train_processed_manual = pd.concat(
    [X_train_num, X_train_cat],
    axis=1
)

X_test_processed_manual = pd.concat(
    [X_test_num, X_test_cat],
    axis=1
)

print("Training shape:", X_train_processed_manual.shape)
print("Testing shape:", X_test_processed_manual.shape)

Training shape: (957, 36)
Testing shape: (240, 36)


In [89]:
X_train_np = X_train_processed_manual.to_numpy()
X_test_np = X_test_processed_manual.to_numpy()

X_train_bias = np.c_[
    np.ones(X_train_np.shape[0]),
    X_train_np
]

X_test_bias = np.c_[
    np.ones(X_test_np.shape[0]),
    X_test_np
]

start_time = time.perf_counter()

beta = np.linalg.inv(
    X_train_bias.T @ X_train_bias
) @ X_train_bias.T @ y_train_manual

train_time_manual_reg = time.perf_counter() - start_time

print("Training shape:", X_train_bias.shape)
print("Number of coefficients:", len(beta))
print("Training time:", train_time_manual_reg, "seconds")

Training shape: (957, 37)
Number of coefficients: 37
Training time: 0.001146699998571421 seconds


In [90]:
start_time = time.perf_counter()

y_manual_pred = X_test_bias @ beta

predict_time_manual_reg = time.perf_counter() - start_time

print("Prediction time:", predict_time_manual_reg, "seconds")

print("\nFirst 10 predictions:")
print(y_manual_pred[:10])

Prediction time: 0.0001350999991700519 seconds

First 10 predictions:
[36.17454485 32.85113528 51.46254424 29.51480927 45.54392984 28.93911736
 37.43041294 36.37540931 30.74769508 43.4610426 ]


In [91]:
start_time = time.perf_counter()

beta = np.linalg.pinv(
    X_train_bias
) @ y_train_manual

train_time_manual_reg = time.perf_counter() - start_time

print("Number of coefficients:", len(beta))
print("Training time:", train_time_manual_reg, "seconds")

Number of coefficients: 37
Training time: 0.01505150000048161 seconds


In [92]:
start_time = time.perf_counter()

y_manual_pred = X_test_bias @ beta

predict_time_manual_reg = time.perf_counter() - start_time

print("Prediction time:", predict_time_manual_reg, "seconds")

print("\nFirst 10 predictions:")
print(y_manual_pred[:10])

Prediction time: 0.0001109000004362315 seconds

First 10 predictions:
[0.63679562 0.73964249 0.81039749 0.72848321 0.76797792 0.79974827
 0.58968705 0.62668732 0.65751647 0.71050959]


In [93]:
errors = y_test_manual - y_manual_pred

mae_manual_reg = np.mean(np.abs(errors))
rmse_manual_reg = np.sqrt(np.mean(errors ** 2))

ss_total = np.sum(
    (y_test_manual - np.mean(y_test_manual)) ** 2
)

ss_residual = np.sum(
    errors ** 2
)

r2_manual_reg = 1 - (ss_residual / ss_total)

print("MAE :", mae_manual_reg)
print("RMSE:", rmse_manual_reg)
print("R2  :", r2_manual_reg)

MAE : 0.11114585853975884
RMSE: 0.14935052152028525
R2  : 0.15994326709578088


In [94]:
y_clf_train_manual = y_clf_train.to_numpy()
y_clf_test_manual = y_clf_test.to_numpy()

print("Training target shape:", y_clf_train_manual.shape)
print("Testing target shape:", y_clf_test_manual.shape)

print("\nFirst 10 training targets:")
print(y_clf_train_manual[:10])

Training target shape: (957,)
Testing target shape: (240,)

First 10 training targets:
[1 0 1 1 1 1 0 1 1 1]


In [95]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))

test_values = np.array([-2, -1, 0, 1, 2])

print(sigmoid(test_values))

[0.11920292 0.26894142 0.5        0.73105858 0.88079708]


In [96]:
weights = np.zeros(X_train_bias.shape[1])

print("Number of weights:", len(weights))
print("First 10 weights:", weights[:10])

Number of weights: 37
First 10 weights: [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [97]:
learning_rate = 0.01
iterations = 1000

start_time = time.perf_counter()

for i in range(iterations):
    z = X_train_bias @ weights
    probabilities = sigmoid(z)

    gradient = (
        X_train_bias.T @ (probabilities - y_clf_train_manual)
    ) / len(y_clf_train_manual)

    weights = weights - learning_rate * gradient

train_time_manual_clf = time.perf_counter() - start_time

print("Training time:", train_time_manual_clf, "seconds")
print("\nFirst 10 weights:")
print(weights[:10])

Training time: 0.04474579999987327 seconds

First 10 weights:
[ 0.51880963  0.00237352 -0.0551088   0.1233633   0.04241411  0.09668269
  0.05447072 -0.27678071 -0.06871485  0.21970819]


In [98]:
start_time = time.perf_counter()

z_test = X_test_bias @ weights
y_clf_prob = sigmoid(z_test)

predict_time_manual_clf = time.perf_counter() - start_time

print("Prediction time:", predict_time_manual_clf, "seconds")

print("\nFirst 10 probabilities:")
print(y_clf_prob[:10])

Prediction time: 0.0002822999995260034 seconds

First 10 probabilities:
[0.7976758  0.81178184 0.6353033  0.58185584 0.59850234 0.80270122
 0.56056448 0.59885496 0.846699   0.80216291]


In [99]:
y_clf_pred_manual = (y_clf_prob >= 0.5).astype(int)

print("First 20 predictions:")
print(y_clf_pred_manual[:20])

print("\nPrediction counts:")
print(pd.Series(y_clf_pred_manual).value_counts())

First 20 predictions:
[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]

Prediction counts:
1    239
0      1
Name: count, dtype: int64


In [100]:
tp = np.sum((y_clf_test_manual == 1) & (y_clf_pred_manual == 1))
tn = np.sum((y_clf_test_manual == 0) & (y_clf_pred_manual == 0))
fp = np.sum((y_clf_test_manual == 0) & (y_clf_pred_manual == 1))
fn = np.sum((y_clf_test_manual == 1) & (y_clf_pred_manual == 0))

accuracy_manual_clf = (tp + tn) / len(y_clf_test_manual)

precision_manual_clf = tp / (tp + fp) if (tp + fp) > 0 else 0
recall_manual_clf = tp / (tp + fn) if (tp + fn) > 0 else 0

f1_manual_clf = (
    2 * precision_manual_clf * recall_manual_clf
    / (precision_manual_clf + recall_manual_clf)
    if (precision_manual_clf + recall_manual_clf) > 0 else 0
)

print("TP:", tp)
print("TN:", tn)
print("FP:", fp)
print("FN:", fn)

print("\nAccuracy :", accuracy_manual_clf)
print("Precision:", precision_manual_clf)
print("Recall   :", recall_manual_clf)
print("F1 Score :", f1_manual_clf)

TP: 177
TN: 1
FP: 62
FN: 0

Accuracy : 0.7416666666666667
Precision: 0.7405857740585774
Recall   : 1.0
F1 Score : 0.8509615384615384


In [101]:
manual_results = {
    "Regression": {
        "MAE": mae_manual_reg,
        "RMSE": rmse_manual_reg,
        "R2": r2_manual_reg,
        "Train Time": train_time_manual_reg,
        "Prediction Time": predict_time_manual_reg
    },
    "Classification": {
        "Accuracy": accuracy_manual_clf,
        "Precision": precision_manual_clf,
        "Recall": recall_manual_clf,
        "F1": f1_manual_clf,
        "Train Time": train_time_manual_clf,
        "Prediction Time": predict_time_manual_clf
    }
}

print(manual_results)

{'Regression': {'MAE': np.float64(0.11114585853975884), 'RMSE': np.float64(0.14935052152028525), 'R2': np.float64(0.15994326709578088), 'Train Time': 0.01505150000048161, 'Prediction Time': 0.0001109000004362315}, 'Classification': {'Accuracy': np.float64(0.7416666666666667), 'Precision': np.float64(0.7405857740585774), 'Recall': np.float64(1.0), 'F1': np.float64(0.8509615384615384), 'Train Time': 0.04474579999987327, 'Prediction Time': 0.0002822999995260034}}


In [102]:
regression_comparison = pd.DataFrame({
    "Metric": ["MAE", "RMSE", "R2", "Train Time", "Prediction Time"],
    "Scikit-learn": [
        sklearn_results["Regression"]["MAE"],
        sklearn_results["Regression"]["RMSE"],
        sklearn_results["Regression"]["R2"],
        sklearn_results["Regression"]["Train Time"],
        sklearn_results["Regression"]["Prediction Time"]
    ],
    "Manual NumPy": [
        manual_results["Regression"]["MAE"],
        manual_results["Regression"]["RMSE"],
        manual_results["Regression"]["R2"],
        manual_results["Regression"]["Train Time"],
        manual_results["Regression"]["Prediction Time"]
    ]
})

print(regression_comparison)

            Metric  Scikit-learn  Manual NumPy
0              MAE      0.111146      0.111146
1             RMSE      0.149351      0.149351
2               R2      0.159943      0.159943
3       Train Time      0.014329      0.015052
4  Prediction Time      0.002126      0.000111


In [103]:
classification_comparison = pd.DataFrame({
    "Metric": ["Accuracy", "Precision", "Recall", "F1", "Train Time", "Prediction Time"],
    "Scikit-learn": [
        sklearn_results["Classification"]["Accuracy"],
        sklearn_results["Classification"]["Precision"],
        sklearn_results["Classification"]["Recall"],
        sklearn_results["Classification"]["F1"],
        sklearn_results["Classification"]["Train Time"],
        sklearn_results["Classification"]["Prediction Time"]
    ],
    "Manual NumPy": [
        manual_results["Classification"]["Accuracy"],
        manual_results["Classification"]["Precision"],
        manual_results["Classification"]["Recall"],
        manual_results["Classification"]["F1"],
        manual_results["Classification"]["Train Time"],
        manual_results["Classification"]["Prediction Time"]
    ]
})

print(classification_comparison)

            Metric  Scikit-learn  Manual NumPy
0         Accuracy      0.766667      0.741667
1        Precision      0.792271      0.740586
2           Recall      0.926554      1.000000
3               F1      0.854167      0.850962
4       Train Time      0.023235      0.044746
5  Prediction Time      0.000544      0.000282


In [104]:
learning_rates = [0.001, 0.005, 0.01, 0.05, 0.1]
iterations = 1000

for lr in learning_rates:
    weights_test = np.zeros(X_train_bias.shape[1])

    for i in range(iterations):
        z = X_train_bias @ weights_test
        probabilities = sigmoid(z)

        gradient = (
            X_train_bias.T @ (probabilities - y_clf_train_manual)
        ) / len(y_clf_train_manual)

        weights_test = weights_test - lr * gradient

    probabilities_test = sigmoid(X_test_bias @ weights_test)
    predictions_test = (probabilities_test >= 0.5).astype(int)

    accuracy = np.mean(predictions_test == y_clf_test_manual)

    print("Learning rate:", lr, "Accuracy:", accuracy)

Learning rate: 0.001 Accuracy: 0.7375
Learning rate: 0.005 Accuracy: 0.7416666666666667
Learning rate: 0.01 Accuracy: 0.7416666666666667
Learning rate: 0.05 Accuracy: 0.7625
Learning rate: 0.1 Accuracy: 0.775


In [105]:
learning_rates = [0.01, 0.05, 0.1]
iterations = 1000

for lr in learning_rates:
    weights_test = np.zeros(X_train_bias.shape[1])

    for i in range(iterations):
        z = X_train_bias @ weights_test
        probabilities = sigmoid(z)

        gradient = (
            X_train_bias.T @ (probabilities - y_clf_train_manual)
        ) / len(y_clf_train_manual)

        weights_test = weights_test - lr * gradient

    probabilities_test = sigmoid(X_test_bias @ weights_test)
    predictions_test = (probabilities_test >= 0.5).astype(int)

    tp = np.sum((y_clf_test_manual == 1) & (predictions_test == 1))
    tn = np.sum((y_clf_test_manual == 0) & (predictions_test == 0))
    fp = np.sum((y_clf_test_manual == 0) & (predictions_test == 1))
    fn = np.sum((y_clf_test_manual == 1) & (predictions_test == 0))

    accuracy = (tp + tn) / len(y_clf_test_manual)
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0
    f1 = (
        2 * precision * recall / (precision + recall)
        if (precision + recall) > 0 else 0
    )

    print("\nLearning rate:", lr)
    print("Accuracy :", accuracy)
    print("Precision:", precision)
    print("Recall   :", recall)
    print("F1 Score :", f1)


Learning rate: 0.01
Accuracy : 0.7416666666666667
Precision: 0.7405857740585774
Recall   : 1.0
F1 Score : 0.8509615384615384

Learning rate: 0.05
Accuracy : 0.7625
Precision: 0.7654867256637168
Recall   : 0.9774011299435028
F1 Score : 0.858560794044665

Learning rate: 0.1
Accuracy : 0.775
Precision: 0.786046511627907
Recall   : 0.9548022598870056
F1 Score : 0.8622448979591836


In [106]:
learning_rate = 0.1
iteration_values = [500, 1000, 2000, 5000]

for iterations in iteration_values:
    weights_test = np.zeros(X_train_bias.shape[1])

    for i in range(iterations):
        z = X_train_bias @ weights_test
        probabilities = sigmoid(z)

        gradient = (
            X_train_bias.T @ (probabilities - y_clf_train_manual)
        ) / len(y_clf_train_manual)

        weights_test = weights_test - learning_rate * gradient

    probabilities_test = sigmoid(X_test_bias @ weights_test)
    predictions_test = (probabilities_test >= 0.5).astype(int)

    tp = np.sum((y_clf_test_manual == 1) & (predictions_test == 1))
    tn = np.sum((y_clf_test_manual == 0) & (predictions_test == 0))
    fp = np.sum((y_clf_test_manual == 0) & (predictions_test == 1))
    fn = np.sum((y_clf_test_manual == 1) & (predictions_test == 0))

    accuracy = (tp + tn) / len(y_clf_test_manual)
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0

    f1 = (
        2 * precision * recall / (precision + recall)
        if (precision + recall) > 0 else 0
    )

    print("\nIterations:", iterations)
    print("Accuracy :", accuracy)
    print("Precision:", precision)
    print("Recall   :", recall)
    print("F1 Score :", f1)


Iterations: 500
Accuracy : 0.7625
Precision: 0.7654867256637168
Recall   : 0.9774011299435028
F1 Score : 0.858560794044665

Iterations: 1000
Accuracy : 0.775
Precision: 0.786046511627907
Recall   : 0.9548022598870056
F1 Score : 0.8622448979591836

Iterations: 2000
Accuracy : 0.7708333333333334
Precision: 0.7932692307692307
Recall   : 0.9322033898305084
F1 Score : 0.8571428571428572

Iterations: 5000
Accuracy : 0.7625
Precision: 0.7912621359223301
Recall   : 0.9209039548022598
F1 Score : 0.8511749347258485


In [107]:
learning_rate = 0.1
iterations = 1000

optimized_weights = np.zeros(X_train_bias.shape[1])

start_time = time.perf_counter()

for i in range(iterations):
    z = X_train_bias @ optimized_weights
    probabilities = sigmoid(z)

    gradient = (
        X_train_bias.T @ (probabilities - y_clf_train_manual)
    ) / len(y_clf_train_manual)

    optimized_weights = optimized_weights - learning_rate * gradient

optimized_train_time = time.perf_counter() - start_time

start_time = time.perf_counter()

optimized_probabilities = sigmoid(X_test_bias @ optimized_weights)
optimized_predictions = (optimized_probabilities >= 0.5).astype(int)

optimized_predict_time = time.perf_counter() - start_time

tp = np.sum((y_clf_test_manual == 1) & (optimized_predictions == 1))
tn = np.sum((y_clf_test_manual == 0) & (optimized_predictions == 0))
fp = np.sum((y_clf_test_manual == 0) & (optimized_predictions == 1))
fn = np.sum((y_clf_test_manual == 1) & (optimized_predictions == 0))

optimized_accuracy = (tp + tn) / len(y_clf_test_manual)
optimized_precision = tp / (tp + fp) if (tp + fp) > 0 else 0
optimized_recall = tp / (tp + fn) if (tp + fn) > 0 else 0

optimized_f1 = (
    2 * optimized_precision * optimized_recall
    / (optimized_precision + optimized_recall)
    if (optimized_precision + optimized_recall) > 0 else 0
)

print("Accuracy :", optimized_accuracy)
print("Precision:", optimized_precision)
print("Recall   :", optimized_recall)
print("F1 Score :", optimized_f1)
print("Train Time:", optimized_train_time)
print("Prediction Time:", optimized_predict_time)

Accuracy : 0.775
Precision: 0.786046511627907
Recall   : 0.9548022598870056
F1 Score : 0.8622448979591836
Train Time: 0.0501098999993701
Prediction Time: 0.00019299999985378236


In [116]:
categorical_train = pd.get_dummies(
    X_train_manual[categorical_cols].astype(str),
    drop_first=False
)

categorical_train = categorical_train.reindex(
    columns=categorical_train.columns
)

feature_names = list(categorical_train.columns) + numerical_cols

print("Feature names:", len(feature_names))
print("Coefficients:", len(beta) - 1)

Feature names: 36
Coefficients: 36


In [117]:
coef_df = pd.DataFrame({
    "Feature": feature_names,
    "Coefficient": beta[1:]
})

coef_df["Abs_Coefficient"] = coef_df["Coefficient"].abs()

display(
    coef_df.sort_values(
        "Abs_Coefficient",
        ascending=False
    ).head(15)
)

,Feature,Coefficient,Abs_Coefficient
17,team_2,0.191266,0.191266
16,team_12,0.189827,0.189827
15,team_11,0.148092,0.148092
8,day_Saturday,0.146526,0.146526
24,team_9,0.101430,0.101430
29,incentive,0.087305,0.087305
1,quarter_Quarter2,-0.086579,0.086579
22,team_7,0.086002,0.086002
19,team_4,0.078429,0.078429
30,idle_time,0.074907,0.074907


In [118]:
thresholds = [0.02, 0.04, 0.06, 0.08]

feature_results = []

for threshold in thresholds:
    selected = coef_df[
        coef_df["Abs_Coefficient"] >= threshold
    ]["Feature"].tolist()

    selected_idx = [
        feature_names.index(f)
        for f in selected
    ]

    X_train_selected = X_train_bias[:, selected_idx]
    X_test_selected = X_test_bias[:, selected_idx]

    beta_selected = np.linalg.pinv(X_train_selected) @ y_train_manual

    y_pred_selected = X_test_selected @ beta_selected

    rmse_selected = np.sqrt(
        np.mean((y_test_manual - y_pred_selected) ** 2)
    )

    feature_results.append({
        "Threshold": threshold,
        "Features": len(selected),
        "RMSE": rmse_selected
    })

pd.DataFrame(feature_results)

,Threshold,Features,RMSE
0,0.02,28,0.150448
1,0.04,24,0.153457
2,0.06,15,0.151282
3,0.08,8,0.479422


In [119]:
print("Full Model")
print("Features:", X_train_bias.shape[1] - 1)
print("RMSE:", rmse_manual_reg)
print("R2:", r2_manual_reg)
print("Train Time:", train_time_manual_reg)
print("Prediction Time:", predict_time_manual_reg)

Full Model
Features: 36
RMSE: 0.14935052152028525
R2: 0.15994326709578088
Train Time: 0.01505150000048161
Prediction Time: 0.0001109000004362315


In [120]:
linear_optimization = pd.DataFrame({
    "Version": [
        "Full Manual Linear Regression",
        "Feature Selection (Threshold 0.02)"
    ],
    "Features": [
        36,
        28
    ],
    "RMSE": [
        rmse_manual_reg,
        feature_results[0]["RMSE"]
    ]
})

display(linear_optimization.round(6))

,Version,Features,RMSE
0,Full Manual Linear Regression,36,0.149351
1,Feature Selection (Threshold 0.02),28,0.150448


In [122]:
optimization_summary = pd.DataFrame({
    "Model": [
        "Manual Linear Regression",
        "Manual Linear Regression",
        "Manual Logistic Regression",
        "Manual Logistic Regression"
    ],
    "Optimization": [
        "Before feature selection",
        "After feature selection test",
        "Before tuning",
        "After learning-rate tuning"
    ],
    "Metric": [
        "RMSE",
        "RMSE",
        "F1 Score",
        "F1 Score"
    ],
    "Score": [
        rmse_manual_reg,
        feature_results[0]["RMSE"],
        manual_results["Classification"]["F1"],
        optimized_f1
    ]
})

display(optimization_summary.round(6))

,Model,Optimization,Metric,Score
0,Manual Linear Regression,Before feature selection,RMSE,0.149351
1,Manual Linear Regression,After feature selection test,RMSE,0.150448
2,Manual Logistic Regression,Before tuning,F1 Score,0.850962
3,Manual Logistic Regression,After learning-rate tuning,F1 Score,0.862245


In [123]:
final_comparison = pd.DataFrame({
    "Model": [
        "Scikit-learn Linear Regression",
        "Manual Linear Regression",
        "Scikit-learn Logistic Regression",
        "Manual Logistic Regression",
        "Optimized Manual Logistic Regression"
    ],
    "Metric": [
        "RMSE",
        "RMSE",
        "F1 Score",
        "F1 Score",
        "F1 Score"
    ],
    "Score": [
        sklearn_results["Regression"]["RMSE"],
        manual_results["Regression"]["RMSE"],
        sklearn_results["Classification"]["F1"],
        manual_results["Classification"]["F1"],
        optimized_f1
    ],
    "Train Time (s)": [
        sklearn_results["Regression"]["Train Time"],
        manual_results["Regression"]["Train Time"],
        sklearn_results["Classification"]["Train Time"],
        manual_results["Classification"]["Train Time"],
        optimized_train_time
    ],
    "Prediction Time (s)": [
        sklearn_results["Regression"]["Prediction Time"],
        manual_results["Regression"]["Prediction Time"],
        sklearn_results["Classification"]["Prediction Time"],
        manual_results["Classification"]["Prediction Time"],
        optimized_predict_time
    ]
})

display(final_comparison.round(6))

,Model,Metric,Score,Train Time (s),Prediction Time (s)
0,Scikit-learn Linear Regression,RMSE,0.149351,0.014329,0.002126
1,Manual Linear Regression,RMSE,0.149351,0.015052,0.000111
2,Scikit-learn Logistic Regression,F1 Score,0.854167,0.023235,0.000544
3,Manual Logistic Regression,F1 Score,0.850962,0.044746,0.000282
4,Optimized Manual Logistic Regression,F1 Score,0.862245,0.050110,0.000193


# **1. Linear Regression:**

### The manual Linear Regression model produced the same RMSE as the Scikit-learn model (0.149351). This shows that the manually implemented normal equation produced equivalent results.

### Feature selection was tested by removing features with smaller coefficient magnitudes. However, the RMSE increased from 0.149351 to 0.150448 after removing features. Therefore, the full 36-feature model was retained.

# **2. Logistic Regression:**

### The initial manual Logistic Regression model achieved an F1 score of 0.850962.

### Different learning rates were tested. Increasing the learning rate from 0.01 to 0.1 improved the F1 score to 0.862245 while keeping the number of iterations at 1000.

### The number of iterations was also tested. Increasing the iterations beyond 1000 did not improve the F1 score, so 1000 iterations was retained.

### Therefore, the optimized manual Logistic Regression used:

### - Learning rate = 0.1
### - Iterations = 1000
### - F1 Score = 0.862245

# **3. Runtime Comparison:**

### The manual Linear Regression training time (0.015052 s) was close to the Scikit-learn training time (0.014329 s).

### The manual Logistic Regression training time was higher than Scikit-learn because the manual implementation uses gradient descent and performs multiple iterations explicitly.

### The optimized manual Logistic Regression took 0.050110 s for training. Its prediction time was 0.000193 s, compared with 0.000544 s for Scikit-learn.

### Runtime differences can depend on the implementation, vectorization, library optimizations, and system conditions.

# **4. Overall Observation:**

### The experiment shows that machine learning algorithms can be implemented using NumPy and Pandas without using ready-made machine learning algorithms.

### The manual Linear Regression implementation matched Scikit-learn closely. Optimization of the manual Logistic Regression through learning-rate tuning improved its F1 score from 0.850962 to 0.862245.

### Vectorized NumPy operations were used throughout the manual implementation to improve computational efficiency.